linalg/amgxsolver.hpp

```diff
@@ -0,0 +1,220 @@
+   AmgXSolver(const MPI_Comm &comm,
+              const std::string &cfgFile);
+
+   /* Constructor for MPI teams (MPI procs share a GPU) */
+   /* nDevs specifies number of devices per node */
+   AmgXSolver(const MPI_Comm &comm,
+              const std::string &cfgFile, int &nDevs);
+
+   void Initialize_ExclusiveGPU(const MPI_Comm &comm, const std::string &cfgFile);
+
+   void Initialize_MPITeams(const MPI_Comm &comm,
+                            const std::string &cfgFile,
+                            const int nDevs);
+
+   void SetMatrix_MPI_GPU_Exclusive(const HypreParMatrix &A,
+                                    const Array<double> &loc_A,
+                                    const Array<int> &loc_I, const Array<int64_t> &loc_J,
+                                    const bool update_mat = false);
+
+   void SetMatrix_MPI_Teams(const HypreParMatrix &A, const Array<double> &loc_A,
+                            const Array<int> &loc_I, const Array<int64_t> &loc_J,
+                            const bool update_mat = false);
+#endif
+
+   virtual void SetOperator(const Operator &op);
+
+   void UpdateOperator(const Operator &op);
+
+   virtual void Mult(const Vector& b, Vector& x) const;
+
+   int GetNumIterations();
+
+   enum AMGX_MODE {SOLVER, PRECONDITIONER};
+
+   void SetMode(AMGX_MODE mode) {amgxMode = mode;}
+
+   ~AmgXSolver();
+
+   void Finalize();
+
+private:
+
+   AMGX_MODE amgxMode = SOLVER;
+
+#ifdef MFEM_USE_MPI
+   //The following methods send vectors to the root node in a MPI team
+   void GatherArray(const Array<double> &inArr, Array<double> &outArr,
+                    const int mpiTeamSz, const MPI_Comm &mpiTeam) const;
+
+   void GatherArray(const Vector &inArr, Vector &outArr,
+                    const int mpiTeamSz, const MPI_Comm &mpiTeam) const;
+
+   void GatherArray(const Array<int> &inArr, Array<int> &outArr,
+                    const int mpiTeamSz, const MPI_Comm &mpiTeam) const;
+
+   void GatherArray(const Array<int64_t> &inArr, Array<int64_t> &outArr,
+                    const int mpiTeamSz, const MPI_Comm &mpiTeam) const;
+
+   //The following methods send vectors to the root node in a MPI team
+   //and store array partitions and displacements
+   void GatherArray(const Vector &inArr, Vector &outArr,
+                    const int mpiTeamSz, const MPI_Comm &mpiTeamComm,
+                    Array<int> &Apart, Array<int> &Adisp) const;
+
+   void ScatterArray(const Vector &inArr, Vector &outArr,
+                     const int mpiTeamSz, const MPI_Comm &mpi_comm,
+                     Array<int> &Apart, Array<int> &Adisp) const;
+
+   void SetMatrix(const HypreParMatrix &A, const bool update_mat = false);
+#endif
+
+   void SetMatrix(const SparseMatrix &A, const bool update_mat = false);
+
+   static int              count;
+
+   // Indicate if this instance has been initialized.
+   bool                    isInitialized = false;
+
+#ifdef MFEM_USE_MPI
```

Comment:
The amount of effort that went into making this work in the multi-GPU setting is impressive!